openadapt/ml/base_provider.py

```diff
@@ -0,0 +1,76 @@
+"""CompletionProvider and related classes."""
+
+from typing import Any
+
+from pydantic import BaseModel
+
+
+class Modality:
+    """LLM Modality class."""
+
+    TEXT = "TEXT"
+    IMAGE = "IMAGE"
+
+
+class Use:
+    """LLM Use Class."""
+
+    COMMERCIAL_USE = "COMMERCIAL"
+    NON_COMMERCIAL_USE = "NON_COMMERCIAL"
+
+
+class Capability:
+    """LLM Capability Class."""
+
+    TRAINING = "TRAINING"
+    TUNING = "TUNING"
+    INFERENCE = "INFERENCE"
+
+
+class Availability:
+    """LLM Availability Class."""
+
+    LOCAL_CPU = "LOCAL_CPU"
+    LOCAL_GPU = "LOCAL_GPU"
+    HOSTED = "HOSTED"
+
+
+class CompletionProvider(BaseModel):
+    """Base CompletionProvider class."""
+
+    Name: str
+    Capabilities: list[Capability]
+    Modalities: list[Modality]
+    Availabilities: list[Availability]
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """Pydantic Config Class."""
+
+        arbitrary_types_allowed = True
+
+    def infer(prompt: str, model_name: str) -> None:
+        """Run inference on the provider."""
+        raise NotImplementedError
+
+    def finetune(prompt_completion_pair: list[dict[str]]) -> None:
+        """Fine-tune the provider."""
+        raise NotImplementedError
+
+    def get_children() -> Any:
+        """Grab all completion provider classes."""
+        return CompletionProvider.__subclasses__()
+
+
+class CompletionProviderFactory:
+    """CompletionProviderFactory class."""
+
+    @staticmethod
+    def get_for_modality(modality: Modality) -> list[CompletionProvider]:
+        """Gets all available CompletionProviders with the given modality."""
+        completion_providers = CompletionProvider.get_children()
```

Comment:
@abrichr On adding ```self``` to the ```CompletionProvider``` class, one would have to pass in parameters to the ```CompletionProvider``` that adhere to its attributes types here inside the factory (else it throws the following error):
```
E   pydantic.error_wrappers.ValidationError: 4 validation errors for CompletionProvider
E   Name
E     field required (type=value_error.missing)
E   Capabilities
E     field required (type=value_error.missing)
E   Modalities
E     field required (type=value_error.missing)
E   Availabilities
E     field required (type=value_error.missing)

pydantic/main.py:341: ValidationError
```

 so we would be generating a ```CompletionProvider```` object with specific parameters. Not sure if that's adheres to: https://www.digitalocean.com/community/tutorials/factory-design-pattern-in-java